README.md
# ripple-address-codec [![NPM](https://img.shields.io/npm/v/ripple-address-codec.svg)](https://npmjs.org/package/ripple-address-codec)

Functions for encoding and decoding XRP Ledger addresses and seeds. Also includes support for encoding/decoding [rippled validator (node) public keys](https://xrpl.org/run-rippled-as-a-validator.html).

## X-address Conversion

All tools and apps in the XRP Ledger ecosystem are encouraged to adopt support for the X-address format. The X-address format is a single Base58 string that encodes an 'Account ID', a (destination) tag, and whether the address is intended for a test network. This prevents users from unintentionally omitting the destination tag when sending and receiving payments and other transactions.

## API

### encodeXAddress(classicAddress: string, tag: number | false, test: boolean): string

Encode a classic address and (optional) tag to an X-address. If `tag` is `false`, the returned X-address explicitly indicates that the recipient does not want a tag to be used. If `test` is `true`, consumers of the address will know that the address is intended for use on test network(s) and the address will start with `T`.

```js
> const api = require('ripple-address-codec')
> api.encodeXAddress('rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf', 4294967295)
'XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8yuPT7y4xaEHi'
```

### decodeXAddress(xAddress: string): {classicAddress: string, tag: number | false, test: boolean}

Decode an X-address to a classic address and tag. If the X-address did not have a tag, the returned object will not have a `tag` field. If the X-address is intended for use on test network(s), `test` will be `true`; if it is intended for use on the main network (mainnet), `test` will be `false`.

```js
> const api = require('ripple-address-codec')
> rippleAddressCodec.decodeXAddress('XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8yuPT7y4xaEHi')
{
  classicAddress: 'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
  tag: 4294967295,
  test: true
}
```

### isValidXAddress(xAddress: string): boolean

Returns `true` if the provided X-address is valid, or `false` otherwise.

```js
> const api = require('ripple-address-codec')
> api.isValidXAddress('XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8yuPT7y4xaEHi')
true
```

### Other functions

```js
> var api = require('ripple-address-codec');
> api.decodeSeed('sEdTM1uX8pu2do5XvTnutH6HsouMaM2')
{ version: [ 1, 225, 75 ],
  bytes: [ 76, 58, 29, 33, 63, 189, 251, 20, 199, 194, 141, 96, 148, 105, 179, 65 ],
  type: 'ed25519' }
> api.decodeSeed('sn259rEFXrQrWyx3Q7XneWcwV6dfL')
{ version: 33,
  bytes: [ 207, 45, 227, 120, 251, 221, 126, 46, 232, 125, 72, 109, 251, 90, 123, 255 ],
  type: 'secp256k1' }
> api.decodeAccountID('rJrRMgiRgrU6hDF4pgu5DXQdWyPbY35ErN')
[ 186,
  142,
  120,
  98,
  110,
  228,
  44,
  65,
  180,
  109,
  70,
  195,
  4,
  141,
  243,
  161,
  195,
  200,
  112,
  114 ]
```

## Tests

Run unit tests with:

    yarn test

Use `--watch` to run in watch mode, so that when you modify the tests, they are automatically re-run:

    yarn test --watch

Use `--coverage` to generate and display code coverage information:

    yarn test --coverage

This tells jest to output code coverage info in the `./coverage` directory, in addition to showing it on the command line.

## Acknowledgements

This library references and adopts code and standards from the following sources:

- [XLS-5d Standard for Tagged Addresses](https://github.com/xrp-community/standards-drafts/issues/6) by @nbougalis
- [XRPL Tagged Address Codec](https://github.com/xrp-community/xrpl-tagged-address-codec) by @WietseWind
- [X-Address transaction functions](https://github.com/codetsunami/xrpl-tools/tree/master/xaddress-functions) by @codetsunami


HISTORY.md
# ripple-address-codec Release History

## 4.0.0 (UNRELEASED)

### Breaking Changes

* `decodeAddress` has been renamed to `decodeAccountID`

### New Features

* `classicAddressToXAddress` - Derive X-address from classic address, tag, and network ID
* `encodeXAddress` - Encode account ID, tag, and network ID as an X-address
* `xAddressToClassicAddress` - Decode an X-address to account ID, tag, and network ID
* `decodeXAddress` - Convert X-address to classic address, tag, and network ID
* `isValidXAddress` - Check whether an X-address (X...) is valid


src/index.test.ts
import {
  classicAddressToXAddress,
  xAddressToClassicAddress,
  isValidXAddress,
  encodeXAddress
} from './index'

const testCases = [
  [
    'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    false,
    'XVLhHMPHU98es4dbozjVtdWzVrDjtV5fdx1mHp98tDMoQXb',
    'TVE26TYGhfLC7tQDno7G8dGtxSkYQn49b3qD26PK7FcGSKE'
  ],
  [
    'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    0,
    'XVLhHMPHU98es4dbozjVtdWzVrDjtV8AqEL4xcZj5whKbmc',
    'TVE26TYGhfLC7tQDno7G8dGtxSkYQnSy8RHqGHoGJ59spi2'
  ],
  [
    'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    1,
    'XVLhHMPHU98es4dbozjVtdWzVrDjtV8xvjGQTYPiAx6gwDC',
    'TVE26TYGhfLC7tQDno7G8dGtxSkYQnSz1uDimDdPYXzSpyw'
  ],
  [
    'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    2,
    'XVLhHMPHU98es4dbozjVtdWzVrDjtV8zpDURx7DzBCkrQE7',
    'TVE26TYGhfLC7tQDno7G8dGtxSkYQnTryP9tG9TW8GeMBmd'
  ],
  [
    'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    32,
    'XVLhHMPHU98es4dbozjVtdWzVrDjtVoYiC9UvKfjKar4LJe',
    'TVE26TYGhfLC7tQDno7G8dGtxSkYQnT2oqaCDzMEuCDAj1j'
  ],
  [
    'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    276,
    'XVLhHMPHU98es4dbozjVtdWzVrDjtVoKj3MnFGMXEFMnvJV',
    'TVE26TYGhfLC7tQDno7G8dGtxSkYQnTMgJJYfAbsiPsc6Zg'
  ],
  [
    'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    65591,
    'XVLhHMPHU98es4dbozjVtdWzVrDjtVozpjdhPQVdt3ghaWw',
    'TVE26TYGhfLC7tQDno7G8dGtxSkYQn7ryu2W6njw7mT1jmS'
  ],
  [
    'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    16781933,
    'XVLhHMPHU98es4dbozjVtdWzVrDjtVqrDUk2vDpkTjPsY73',
    'TVE26TYGhfLC7tQDno7G8dGtxSkYQnVsw45sDtGHhLi27Qa'
  ],
  [
    'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    4294967294,
    'XVLhHMPHU98es4dbozjVtdWzVrDjtV1kAsixQTdMjbWi39u',
    'TVE26TYGhfLC7tQDno7G8dGtxSkYQnX8tDFQ53itLNqs6vU'
  ],
  [
    'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    4294967295,
    'XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8yuPT7y4xaEHi',
    'TVE26TYGhfLC7tQDno7G8dGtxSkYQnXoy6kSDh6rZzApc69'
  ],
  [
    'rPEPPER7kfTD9w2To4CQk6UCfuHM9c6GDY',
    false,
    'XV5sbjUmgPpvXv4ixFWZ5ptAYZ6PD2gYsjNFQLKYW33DzBm',
    'TVd2rqMkYL2AyS97NdELcpeiprNBjwLZzuUG5rZnaewsahi'
  ],
  [
    'rPEPPER7kfTD9w2To4CQk6UCfuHM9c6GDY',
    0,
    'XV5sbjUmgPpvXv4ixFWZ5ptAYZ6PD2m4Er6SnvjVLpMWPjR',
    'TVd2rqMkYL2AyS97NdELcpeiprNBjwRQUBetPbyrvXSTuxU'
  ],
  [
    'rPEPPER7kfTD9w2To4CQk6UCfuHM9c6GDY',
    13371337,
    'XV5sbjUmgPpvXv4ixFWZ5ptAYZ6PD2qwGkhgc48zzcx6Gkr',
    'TVd2rqMkYL2AyS97NdELcpeiprNBjwVUDvp3vhpXbNhLwJi'
  ]
]

;[false, true].forEach(isTestAddress => {
  const MAX_32_BIT_UNSIGNED_INT = 4294967295
  const network = isTestAddress ? ' (test)' : ' (main)'

  for (const i in testCases) {
    const testCase = testCases[i]
    const classicAddress = testCase[0] as string
    const tag = testCase[1] !== false ? testCase[1] as number : false
    const xAddress = isTestAddress ? testCase[3] as string : testCase[2] as string
    test(`Converts ${classicAddress}${tag ? ':' + tag : ''} to ${xAddress}${network}`, () => {
      expect(classicAddressToXAddress(classicAddress, tag, isTestAddress)).toBe(xAddress)
      const myClassicAddress = xAddressToClassicAddress(xAddress)
      expect(myClassicAddress).toEqual({
        classicAddress,
        tag,
        test: isTestAddress
      })
      expect(isValidXAddress(xAddress)).toBe(true)
    })
  }

  {
    const classicAddress = 'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf'
    const tag = MAX_32_BIT_UNSIGNED_INT + 1

    test(`Converting ${classicAddress}:${tag}${network} throws`, () => {
      expect(() => {
        classicAddressToXAddress(classicAddress, tag, isTestAddress)
      }).toThrowError(new Error('Invalid tag'))
    })
  }

  {
    const classicAddress = 'r'
    test(`Invalid classic address: Converting ${classicAddress}${network} throws`, () => {
      expect(() => {
        classicAddressToXAddress(classicAddress, false, isTestAddress)
      }).toThrowError(new Error('invalid_input_size: decoded data must have length >= 5'))
    })
  }

  {
    const highAndLowAccounts = [
      Buffer.from('00'.repeat(20), 'hex'),
      Buffer.from('00'.repeat(19) + '01', 'hex'),
      Buffer.from('01'.repeat(20), 'hex'),
      Buffer.from('FF'.repeat(20), 'hex')
    ]

    highAndLowAccounts.forEach(accountId => {
      [false, 0, 1, MAX_32_BIT_UNSIGNED_INT].forEach(t => {
        const tag = (t as number | false)
        const xAddress = encodeXAddress(accountId, tag, isTestAddress)
        test(`Encoding ${accountId.toString('hex')}${tag ? ':' + tag : ''} to ${xAddress} has expected length`, () => {
          expect(xAddress.length).toBe(47)
        })
      })
    })
  }
})

{
  const xAddress = 'XVLhHMPHU98es4dbozjVtdWzVrDjtV5fdx1mHp98tDMoQXa'
  test(`Invalid X-address (bad checksum): Converting ${xAddress} throws`, () => {
    expect(() => {
      xAddressToClassicAddress(xAddress)
    }).toThrowError(new Error('checksum_invalid'))
  })
}

{
  const xAddress = 'dGzKGt8CVpWoa8aWL1k18tAdy9Won3PxynvbbpkAqp3V47g'
  test(`Invalid X-address (bad prefix): Converting ${xAddress} throws`, () => {
    expect(() => {
      xAddressToClassicAddress(xAddress)
    }).toThrowError(new Error('Invalid X-address: bad prefix'))
  })
}

test(`Invalid X-address (64-bit tag) throws`, () => {
  expect(() => {
    // Encoded from:
    // {
    //   classicAddress: 'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
    //   tag: MAX_32_BIT_UNSIGNED_INT + 1
    // }
    xAddressToClassicAddress('XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8zeUygYrCgrPh')
  }).toThrowError('Unsupported X-address')
})

test(`Invalid Account ID throws`, () => {
  expect(() => {
    encodeXAddress(Buffer.from('00'.repeat(19), 'hex'), false, false)
  }).toThrowError('Account ID must be 20 bytes')
})

test(`isValidXAddress returns false for invalid X-address`, () => {
  expect(isValidXAddress('XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8zeUygYrCgrPh')).toBe(false)
})


test/codec-test.js
/* eslint-disable no-unused-expressions/no-unused-expressions */

'use strict'

const assert = require('assert')
const api = require('../')

function toHex(bytes) {
  return new Buffer(bytes).toString('hex').toUpperCase()
}

function toBytes(hex) {
  return new Buffer(hex, 'hex').toJSON().data
}

function uInt32LE_ToUInt32(hex) {
  return Buffer.from(hex, 'hex').readUInt32LE()
}

function uInt32_ToUInt32LE(int) {
  const buf = new Buffer(8)
  buf.writeUInt32LE(int, 0)
  return buf.toString('hex').toUpperCase()
}

describe('ripple-address-codec', function() {
  function makeTest(type, base58, hex) {
    it('can translate between ' + hex + ' and ' + base58, function() {
      const actual = api['encode' + type](toBytes(hex))
      assert.equal(actual, base58)
    })
    it('can translate between ' + base58 + ' and ' + hex, function() {
      const buf = api['decode' + type](base58)
      assert.equal(toHex(buf), hex)
    })
  }

  makeTest('AccountID', 'rJrRMgiRgrU6hDF4pgu5DXQdWyPbY35ErN',
    'BA8E78626EE42C41B46D46C3048DF3A1C3C87072')

  makeTest(
    'NodePublic',
    'n9MXXueo837zYH36DvMc13BwHcqtfAWNJY5czWVbp7uYTj7x17TH',
    '0388E5BA87A000CB807240DF8C848EB0B5FFA5C8E5A521BC8E105C0F0A44217828')

  makeTest('K256Seed', 'sn259rEFXrQrWyx3Q7XneWcwV6dfL',
    'CF2DE378FBDD7E2EE87D486DFB5A7BFF')

  makeTest('EdSeed', 'sEdTM1uX8pu2do5XvTnutH6HsouMaM2',
    '4C3A1D213FBDFB14C7C28D609469B341')

  it('can decode arbitray seeds', function() {
    const decoded = api.decodeSeed('sEdTM1uX8pu2do5XvTnutH6HsouMaM2')
    assert.equal(toHex(decoded.bytes), '4C3A1D213FBDFB14C7C28D609469B341')
    assert.equal(decoded.type, 'ed25519')

    const decoded2 = api.decodeSeed('sn259rEFXrQrWyx3Q7XneWcwV6dfL')
    assert.equal(toHex(decoded2.bytes), 'CF2DE378FBDD7E2EE87D486DFB5A7BFF')
    assert.equal(decoded2.type, 'secp256k1')
  })

  it('can pass a type as second arg to encodeSeed', function() {
    const edSeed = 'sEdTM1uX8pu2do5XvTnutH6HsouMaM2'
    const decoded = api.decodeSeed(edSeed)
    assert.equal(toHex(decoded.bytes), '4C3A1D213FBDFB14C7C28D609469B341')
    assert.equal(decoded.type, 'ed25519')
    assert.equal(api.encodeSeed(decoded.bytes, decoded.type), edSeed)
  })

  it('can decode a tagged address (XLS-5d) to AccountID and tag', function() {
    const tagged = 'r1WTvVjuoBM9vsm2p395AyzCQcJyE9CCWEMvo5U5sP5Px5'
    const decodedHex = toHex(api.decodeTaggedAddress(tagged))
    const encoded = api.encodeAccountID(toBytes(decodedHex.slice(0, -18)))
    assert.equal(encoded, 'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf')
    assert.equal(decodedHex.slice(-18, -16), '01')
    assert.equal(uInt32LE_ToUInt32(decodedHex.slice(-16)), 16781933)
  })
  it('can encode AccountID without tag to tagged address (XLS-5d)', function() {
    const account = 'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf'
    const decoded = toHex(api.decodeAccountID(account))
    const taggedHex = toBytes(decoded + '00' + uInt32_ToUInt32LE(0))
    const encoded = api.encodeTaggedAddress(taggedHex)
    assert.equal(encoded, 'r1WTvVjuoBM9vsm2p395AyzCQcJyEp8aG4YHcqE3XLDehK')
  })
  it('can encode AccountID + tag to tagged address (XLS-5d)', function() {
    const account = 'rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf'
    const tag = 4294967294
    const decoded = toHex(api.decodeAccountID(account))
    const taggedHex = toBytes(decoded + '01' + uInt32_ToUInt32LE(tag))
    const encoded = api.encodeTaggedAddress(taggedHex)
    assert.equal(encoded, 'r1WTvVjuoBM9vsm2p395AyzCQcJyEURPMMjRhJoyxQhdt5')
  })

  it('isValidAddress - secp256k1 address valid', function() {
    assert(api.isValidAddress('rU6K7V3Po4snVhBBaU29sesqs2qTQJWDw1'))
  })
  it('isValidAddress - ed25519 address valid', function() {
    assert(api.isValidAddress('rLUEXYuLiQptky37CqLcm9USQpPiz5rkpD'))
  })
  it('isValidAddress - invalid', function() {
    assert(!api.isValidAddress('rU6K7V3Po4snVhBBaU29sesqs2qTQJWDw2'))
  })
  it('isValidAddress - empty', function() {
    assert(!api.isValidAddress(''))
  })
  it('isValidTaggedAddress - secp256k1 tagged address valid', function() {
    const address = 'r1WTvVjuoBM9vsm2p395AyzCQcJyE3eVsfqQrBa3X4q4qF'
    assert(api.isValidTaggedAddress(address))
  })
  it('isValidTaggedAddress - invalid for untagged', function() {
    assert(!api.isValidTaggedAddress('rGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf'))
  })

})


src/xrp-codec.test.ts
import * as api from './xrp-codec'

function toHex(bytes: Buffer) {
  return Buffer.from(bytes).toString('hex').toUpperCase()
}

function toBytes(hex: string) {
  return Buffer.from(hex, 'hex').toJSON().data
}

function makeTest(encoder: Function, decoder: Function, base58: string, hex: string) {
  test(`can translate between ${hex} and ${base58}`, function() {
    const actual = encoder(toBytes(hex))
    expect(actual).toBe(base58)
  })
  test(`can translate between ${base58} and ${hex})`, function() {
    const buf = decoder(base58)
    expect(toHex(buf)).toBe(hex)
  })
}

makeTest(api.encodeAccountID, api.decodeAccountID, 'rJrRMgiRgrU6hDF4pgu5DXQdWyPbY35ErN',
  'BA8E78626EE42C41B46D46C3048DF3A1C3C87072')

makeTest(api.encodeNodePublic, api.decodeNodePublic,
  'n9MXXueo837zYH36DvMc13BwHcqtfAWNJY5czWVbp7uYTj7x17TH',
  '0388E5BA87A000CB807240DF8C848EB0B5FFA5C8E5A521BC8E105C0F0A44217828')

test('can decode arbitrary seeds', function() {
  const decoded = api.decodeSeed('sEdTM1uX8pu2do5XvTnutH6HsouMaM2')
  expect(toHex(decoded.bytes)).toBe('4C3A1D213FBDFB14C7C28D609469B341')
  expect(decoded.type).toBe('ed25519')

  const decoded2 = api.decodeSeed('sn259rEFXrQrWyx3Q7XneWcwV6dfL')
  expect(toHex(decoded2.bytes)).toBe('CF2DE378FBDD7E2EE87D486DFB5A7BFF')
  expect(decoded2.type).toBe('secp256k1')
})

test('can pass a type as second arg to encodeSeed', function() {
  const edSeed = 'sEdTM1uX8pu2do5XvTnutH6HsouMaM2'
  const decoded = api.decodeSeed(edSeed)
  expect(toHex(decoded.bytes)).toBe('4C3A1D213FBDFB14C7C28D609469B341')
  expect(decoded.type).toBe('ed25519')
  expect(api.encodeSeed(decoded.bytes, /* decoded.type */ 'ed25519')).toBe(edSeed)
})

test('isValidAddress - secp256k1 address valid', function() {
  expect(api.isValidAddress('rU6K7V3Po4snVhBBaU29sesqs2qTQJWDw1')).toBe(true)
})

test('isValidAddress - ed25519 address valid', function() {
  expect(api.isValidAddress('rLUEXYuLiQptky37CqLcm9USQpPiz5rkpD')).toBe(true)
})

test('isValidAddress - invalid', function() {
  expect(api.isValidAddress('rU6K7V3Po4snVhBBaU29sesqs2qTQJWDw2')).toBe(false)
})

test('isValidAddress - empty', function() {
  expect(api.isValidAddress('')).toBe(false)
})

describe('encodeSeed', function() {

  it('encodes a secp256k1 seed', function() {
    const result = api.encodeSeed(Buffer.from('CF2DE378FBDD7E2EE87D486DFB5A7BFF', 'hex'), 'secp256k1')
    expect(result).toBe('sn259rEFXrQrWyx3Q7XneWcwV6dfL')
  })

  it('encodes low secp256k1 seed', function() {
    const result = api.encodeSeed(Buffer.from('00000000000000000000000000000000', 'hex'), 'secp256k1')
    expect(result).toBe('sp6JS7f14BuwFY8Mw6bTtLKWauoUs')
  })

  it('encodes high secp256k1 seed', function() {
    const result = api.encodeSeed(Buffer.from('FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF', 'hex'), 'secp256k1')
    expect(result).toBe('saGwBRReqUNKuWNLpUAq8i8NkXEPN')
  })

  it('encodes an ed25519 seed', function() {
    const result = api.encodeSeed(Buffer.from('4C3A1D213FBDFB14C7C28D609469B341', 'hex'), 'ed25519')
    expect(result).toBe('sEdTM1uX8pu2do5XvTnutH6HsouMaM2')
  })

  it('encodes low ed25519 seed', function() {
    const result = api.encodeSeed(Buffer.from('00000000000000000000000000000000', 'hex'), 'ed25519')
    expect(result).toBe('sEdSJHS4oiAdz7w2X2ni1gFiqtbJHqE')
  })

  it('encodes high ed25519 seed', function() {
    const result = api.encodeSeed(Buffer.from('FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF', 'hex'), 'ed25519')
    expect(result).toBe('sEdV19BLfeQeKdEXyYA4NhjPJe6XBfG')
  })

  test('attempting to encode a seed with less than 16 bytes of entropy throws', function() {
    expect(() => {
      api.encodeSeed(Buffer.from('CF2DE378FBDD7E2EE87D486DFB5A7B', 'hex'), 'secp256k1')
    }).toThrow('entropy must have length 16')
  })

  test('attempting to encode a seed with more than 16 bytes of entropy throws', function() {
    expect(() => {
      api.encodeSeed(Buffer.from('CF2DE378FBDD7E2EE87D486DFB5A7BFFFF', 'hex'), 'secp256k1')
    }).toThrow('entropy must have length 16')
  })
})

describe('decodeSeed', function() {

  it('can decode an Ed25519 seed', function() {
    const decoded = api.decodeSeed('sEdTM1uX8pu2do5XvTnutH6HsouMaM2')
    expect(toHex(decoded.bytes)).toBe('4C3A1D213FBDFB14C7C28D609469B341')
    expect(decoded.type).toBe('ed25519')
  })

  it('can decode a secp256k1 seed', function() {
    const decoded = api.decodeSeed('sn259rEFXrQrWyx3Q7XneWcwV6dfL')
    expect(toHex(decoded.bytes)).toBe('CF2DE378FBDD7E2EE87D486DFB5A7BFF')
    expect(decoded.type).toBe('secp256k1')
  })
})

describe('encodeAccountID', function() {

  it('can encode an AccountID', function() {
    const encoded = api.encodeAccountID(Buffer.from('BA8E78626EE42C41B46D46C3048DF3A1C3C87072', 'hex'))
    expect(encoded).toBe('rJrRMgiRgrU6hDF4pgu5DXQdWyPbY35ErN')
  })

  test('unexpected length should throw', function() {
    expect(() => {
      api.encodeAccountID(Buffer.from('ABCDEF', 'hex'))
    }).toThrow(
      'unexpected_payload_length: bytes.length does not match expectedLength'
    )
  })
})

describe('decodeNodePublic', function() {

  it('can decode a NodePublic', function() {
    const decoded = api.decodeNodePublic('n9MXXueo837zYH36DvMc13BwHcqtfAWNJY5czWVbp7uYTj7x17TH')
    expect(toHex(decoded)).toBe('0388E5BA87A000CB807240DF8C848EB0B5FFA5C8E5A521BC8E105C0F0A44217828')
  })
})

test('encodes 123456789 with version byte of 0', () => {
  expect(api.codec.encode(Buffer.from('123456789'), {
    versions: [0],
    expectedLength: 9
  })).toBe('rnaC7gW34M77Kneb78s')
})

test('multiple versions with no expected length should throw', () => {
  expect(() => {
    api.codec.decode('rnaC7gW34M77Kneb78s', {
      versions: [0, 1]
    })
  }).toThrow('expectedLength is required because there are >= 2 possible versions')
})

test('attempting to decode data with length < 5 should throw', () => {
  expect(() => {
    api.codec.decode('1234', {
      versions: [0]
    })
  }).toThrow('invalid_input_size: decoded data must have length >= 5')
})

test('attempting to decode data with unexpected version should throw', () => {
  expect(() => {
    api.codec.decode('rnaC7gW34M77Kneb78s', {
      versions: [2]
    })
  }).toThrow('version_invalid: version bytes do not match any of the provided version(s)')
})

test('invalid checksum should throw', () => {
  expect(() => {
    api.codec.decode('123456789', {
      versions: [0, 1]
    })
  }).toThrow('checksum_invalid')
})

test('empty payload should throw', () => {
  expect(() => {
    api.codec.decode('', {
      versions: [0, 1]
    })
  }).toThrow('invalid_input_size: decoded data must have length >= 5')
})

test('decode data', () => {
  expect(api.codec.decode('rnaC7gW34M77Kneb78s', {
    versions: [0]
  })).toStrictEqual({
    version: [0],
    bytes: Buffer.from('123456789'),
    type: null
  })
})

test('decode data with expected length', function() {
  expect(api.codec.decode('rnaC7gW34M77Kneb78s', {
      versions: [0],
      expectedLength: 9
    })
    ).toStrictEqual({
      version: [0],
      bytes: Buffer.from('123456789'),
      type: null
    })
})

test('decode data with wrong expected length should throw', function() {
  expect(() => {
    api.codec.decode('rnaC7gW34M77Kneb78s', {
      versions: [0],
      expectedLength: 8
    })
  }).toThrow(
    'version_invalid: version bytes do not match any of the provided version(s)'
  )
  expect(() => {
    api.codec.decode('rnaC7gW34M77Kneb78s', {
      versions: [0],
      expectedLength: 10
    })
  }).toThrow(
    'version_invalid: version bytes do not match any of the provided version(s)'
  )
})


src/index.js
'use strict'

const createHash = require('create-hash')
const apiFactory = require('x-address-codec')

const NODE_PUBLIC = 28
const NODE_PRIVATE = 32
const ACCOUNT_ID = 0
const FAMILY_SEED = 33
const ED25519_SEED = [0x01, 0xE1, 0x4B]

module.exports = apiFactory({
  sha256: function(bytes) {
    return createHash('sha256').update(new Buffer(bytes)).digest()
  },
  defaultAlphabet: 'ripple',
  codecMethods: {
    EdSeed: {
      expectedLength: 16,
      version: ED25519_SEED
    },
    Seed: {
      // TODO: Use a map, not a parallel array
      versionTypes: ['ed25519', 'secp256k1'],
      versions: [ED25519_SEED, FAMILY_SEED],
      expectedLength: 16
    },
    AccountID: {version: ACCOUNT_ID, expectedLength: 20},
    Address: {version: ACCOUNT_ID, expectedLength: 20},
    TaggedAddress: {version: ACCOUNT_ID, expectedLength: 29},
    NodePublic: {version: NODE_PUBLIC, expectedLength: 33},
    NodePrivate: {version: NODE_PRIVATE, expectedLength: 32},
    K256Seed: {version: FAMILY_SEED, expectedLength: 16}}
})


src/utils.ts
// Utilities

type Sequence = number[] | Buffer | Uint8Array

/**
 * Check whether two sequences (e.g. arrays of numbers) are equal.
 *
 * @param arr1 One of the arrays to compare.
 * @param arr2 The other array to compare.
 */
export function seqEqual(arr1: Sequence, arr2: Sequence): boolean {
  if (arr1.length !== arr2.length) {
    return false
  }

  for (let i = 0; i < arr1.length; i++) {
    if (arr1[i] !== arr2[i]) {
      return false
    }
  }
  return true
}

/**
 * Check whether a value is a sequence (e.g. array of numbers).
 *
 * @param val The value to check.
 */
function isSequence(val: Sequence | number): val is Sequence {
  return (val as Sequence).length !== undefined
}

/**
* Concatenate all `arguments` into a single array. Each argument can be either
* a single element or a sequence, which has a `length` property and supports
* element retrieval via sequence[ix].
*
* > concatArgs(1, [2, 3], Buffer.from([4,5]), new Uint8Array([6, 7]));
*  [1,2,3,4,5,6,7]
*
* @returns {number[]} Array of concatenated arguments
*/
export function concatArgs(...myArgs: (number | Sequence)[]): number[] {
  const ret: number[] = []

  myArgs.forEach(function (arg) {
    if (isSequence(arg)) {
      for (let j = 0; j < arg.length; j++) {
        ret.push(arg[j])
      }
    } else {
      ret.push(arg)
    }
  })
  return ret
}


src/index.ts
import {
  codec,
  encodeSeed,
  decodeSeed,
  encodeAccountID,
  decodeAccountID,
  encodeNodePublic,
  decodeNodePublic,
  isValidAddress
} from './xrp-codec'
import * as assert from 'assert'

const PREFIX_BYTES = {
  MAIN: Buffer.from([0x05, 0x44]), // 5, 68
  TEST: Buffer.from([0x04, 0x93]) // 4, 147
}

function classicAddressToXAddress(classicAddress: string, tag: number | false, test: boolean): string {
  const accountId = decodeAccountID(classicAddress)
  return encodeXAddress(accountId, tag, test)
}

function encodeXAddress(accountId: Buffer, tag: number | false, test: boolean): string {
  if (accountId.length !== 20) {
    // RIPEMD160 is 160 bits = 20 bytes
    throw new Error('Account ID must be 20 bytes')
  }
  const MAX_32_BIT_UNSIGNED_INT = 4294967295
  const flag = tag === false ? 0 : tag <= MAX_32_BIT_UNSIGNED_INT ? 1 : 2
  if (flag === 2) {
    throw new Error('Invalid tag')
  }
  if (tag === false) {
    tag = 0
  }
  const bytes = Buffer.concat(
    [
      test ? PREFIX_BYTES.TEST : PREFIX_BYTES.MAIN,
      accountId,
      Buffer.from(
        [
          flag, // 0x00 if no tag, 0x01 if 32-bit tag
          tag & 0xff, // first byte
          (tag >> 8) & 0xff, // second byte
          (tag >> 16) & 0xff, // third byte
          (tag >> 24) & 0xff, // fourth byte
          0, 0, 0, 0 // four zero bytes (reserved for 64-bit tags)
        ]
      )
    ]
  )
  const xAddress = codec.encodeChecked(bytes)
  return xAddress
}

function xAddressToClassicAddress(xAddress: string): {classicAddress: string, tag: number | false, test: boolean} {
  const {
    accountId,
    tag,
    test
  } = decodeXAddress(xAddress)
  const classicAddress = encodeAccountID(accountId)
  return {
    classicAddress,
    tag,
    test
  }
}

function decodeXAddress(xAddress: string): {accountId: Buffer, tag: number | false, test: boolean} {
  const decoded = codec.decodeChecked(xAddress)
  const test = isBufferForTestAddress(decoded)
  const accountId = decoded.slice(2, 22)
  const tag = tagFromBuffer(decoded)
  return {
    accountId,
    tag,
    test
  }
}

function isBufferForTestAddress(buf: Buffer): boolean {
  const decodedPrefix = buf.slice(0, 2)
  if (PREFIX_BYTES.MAIN.equals(decodedPrefix)) {
    return false
  } else if (PREFIX_BYTES.TEST.equals(decodedPrefix)) {
    return true
  } else {
    throw new Error('Invalid X-address: bad prefix')
  }
}

function tagFromBuffer(buf: Buffer): number | false {
  const flag = buf[22]
  if (flag >= 2) {
    // No support for 64-bit tags at this time
    throw new Error('Unsupported X-address')
  }
  if (flag === 1) {
    // Little-endian to big-endian
    return buf[23] + buf[24] * 0x100 + buf[25] * 0x10000 + buf[26] * 0x1000000
  }
  assert.strictEqual(flag, 0, 'flag must be zero to indicate no tag')
  assert.ok(Buffer.from('0000000000000000', 'hex').equals(buf.slice(23, 23 + 8)),
    'remaining bytes must be zero')
  return false
}

function isValidXAddress(xAddress: string): boolean {
  try {
    decodeXAddress(xAddress)
  } catch (e) {
    return false
  }
  return true
}

export {
  codec, // Codec with XRP alphabet
  encodeSeed, // Encode entropy as a "seed"
  decodeSeed, // Decode a seed into an object with its version, type, and bytes
  encodeAccountID, // Encode bytes as a classic address (r...)
  decodeAccountID, // Decode a classic address to its raw bytes
  encodeNodePublic, // Encode bytes to XRP Ledger node public key format
  decodeNodePublic, // Decode an XRP Ledger node public key into its raw bytes
  isValidAddress, // Check whether a classic address (r...) is valid
  classicAddressToXAddress, // Derive X-address from classic address, tag, and network ID
  encodeXAddress, // Encode account ID, tag, and network ID to X-address
  xAddressToClassicAddress, // Decode X-address to account ID, tag, and network ID
  decodeXAddress, // Convert X-address to classic address, tag, and network ID
  isValidXAddress // Check whether an X-address (X...) is valid
}
